---
title: "HiTIMED"
author: "Ze Zhang"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
    %\VignetteIndexEntry{HiTIMED}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(cache = FALSE, warning = FALSE, message = FALSE, 
                        cache.lazy = FALSE,collapse = TRUE, comment = "#>"
)
library(HiTIMED)
```

The HiTIMED package contains reference libraries derived from 
Illumina HumanMethylation450K DNA methylation 
microarrays (Zhang Z, Salas LA et al. 2022).

The reference libraries were used to estimate proportions of 17 cell types 
(tumor, epithelial, endothelial, stromal, basophil, eosinophil, neutrophil, 
dendritic cell, monocyte, B naïve, B memory, CD4T naïve, CD4T memory,
CD8T naïve, CD8T memory, T regulatory, and natural killer cells) in 3 major 
tumor microenvironment components (tumor, immune, angiogenic) for 20 types 
of carcinomas using a modified version of the algorithm constrained 
projection/quadratic programming described in Houseman et al. 2012. 

**Objects included:**  
1. *HiTIMED_Library* contains 240 libraries for deconvolution  
```{r eval=FALSE}
data("HiTIMED_Library") 
```  

2. *Example_Beta* contains a methylation beta matrix for 3 colon adenocarcinoma
samples from TCGA for HiTIMED function testing.
```{r eval=FALSE}
data("Example_Beta.rda") 
```  

**HiTIMED_deconvolution function for tumor microenvironment deconvolution:**  
 
 We offer the function HiTIMED_deconvolution to estimate proportions for 17 
 cell types (tumor, epithelial, endothelial, stromal, basophil, eosinophil, 
 neutrophil, dendritic cell, monocyte, B naïve, B memory, CD4T naïve, CD4T 
 memory, CD8T naïve, CD8T memory, T regulatory, and natural killer cells) in 
 3 major tumor microenvironment components (tumor, immune, angiogenic) for 20 
 types of carcinomas using a modified version of the algorithm constrained 
 projection/quadratic programming described in Houseman et al. 2012. 
 *see ?HiTIMED_deconvolution for details*  
```{r eval=TRUE}
# Step 1: Load example data
data("Example_Beta") 
# Note: If your machine does not allow access to internet you can download
# and save the file. Then load it manually into the R environment


# Step 2: Load the library
data("HiTIMED_Library")


# Step 3: Run HiTIMED and show results
HiTIMED_result<-HiTIMED_deconvolution(Example_Beta,"COAD",6,"tumor")
head(HiTIMED_result)


```

```{r}
sessionInfo()
```


**References**

Z Zhang, LA Salas et al. (2023) HiTIMED: hierarchical tumor immune 
microenvironment epigenetic deconvolution for accurate cell type resolution 
in the tumor microenvironment using tumor-type-specific DNA methylation data.
Journal of Tranlsational Medicine, 8;20(1):516. 
doi: [10.1186/s12967-022-03736-6] 
(https://dx.doi.org/10.1186/s12967-022-03736-6).


Zheng X et al. (2017). Estimating and accounting for tumor 
purity in the analysis of DNA methylation data from cancer studies. Genome Biol. 
2017;18(1):17. doi: [10.1186/s13059-016-1143-5]
(https://doi.org/10.1186/s13059-016-1143-5). 


LA Salas et al. (2018). An optimized library for reference-based deconvolution
of whole-blood biospecimens assayed using the Illumina HumanMethylationEPIC 
BeadArray. Genome Biology 19, 64. doi: [10.1186/s13059-018-1448-7]
(https://dx.doi.org/10.1186/s13059-018-1448-7).


DC Koestler et al. (2016). Improving cell mixture deconvolution
by identifying optimal DNA methylation libraries (IDOL)}. BMC bioinformatics.
doi: [10.1186/s12859-016-0943-7]
(https://dx.doi.org/10.1186/s12859-016-0943-7)


LA Salas et al. (2022). Enhanced cell deconvolution of peripheral blood using 
DNA methylation for high-resolution immune profiling. 
Nature Communications 13(1):761. 
doi:[10.1038/s41467-021-27864-7](https://dx.doi.org/10.1038/s41467-021-27864-7).  


EA Houseman et al. (2012) DNA methylation arrays as surrogate
measures of cell mixture distribution. BMC Bioinformatics 13, 86.
doi: [10.1186/1471-2105-13-86](https://dx.doi.org/10.1186/1471-2105-13-86).  

